src/coreclr/vm/syncblk.inl

```diff
@@ -74,8 +74,127 @@ FORCEINLINE ObjHeader::HeaderLockResult ObjHeader::AcquireHeaderThinLock(DWORD t
+    // Thin lock is an optimization for such scenarios.
+    //
+    // If we see a thin lock held by other thread for longer than ~5 microseconds, we will "inflate" the lock
+    // and let the adaptive spinning in the fat Lock sort it out whether we have a contentious lock or a long-held lock.
+    //
+    // Another thing to consider is that SpinWait(1) is calibrated to about 35-50 nanoseconds.
+    // It can take much longer only if nop/pause takes much longer, which it should not, as that would be getting
+    // close to the RAM latency.
+    //
+    // Considering that taking and releasing the lock takes 2 CAS instructions + some overhead, we can estimate shortest
+    // time the lock can be held to be in hundreds of nanoseconds. Thus it is unlikely to see more than
+    // 8-10 threads contending for the lock without inflating it. Therefore we can expect to acquire a thin lock in
+    // under 16 tries.
+    //
+    // As for the backoff strategy we have two choices:
+    // Exponential back-off with a limit:
+    //   0, 1, 2, 4, 8, 8, 8, 8, 8, 8, 8, . . . .
+    //
+    // Linear back-off
+    //   0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, . . . .
+    //
+    // In this case these strategies are close in terms of average and worst case latency, so we will prefer linear
+    // back-off as it favors micro-contention scenario, which we expect.
+    //
+
+    for (int i = 0; i < 16; ++i)
+    {
+        LONG oldValue = m_SyncBlockValue.LoadWithoutBarrier();
+
+        if ((oldValue & (BIT_SBLK_IS_HASH_OR_SYNCBLKINDEX +
+                        BIT_SBLK_SPIN_LOCK +
+                        SBLK_MASK_LOCK_THREADID +
+                        SBLK_MASK_LOCK_RECLEVEL)) == 0)
+        {
+            LONG newValue = oldValue | tid;
+#if defined(TARGET_WINDOWS) && defined(TARGET_ARM64)
+            if (FastInterlockedCompareExchangeAcquire((LONG*)&m_SyncBlockValue, newValue, oldValue) == oldValue)
+#else
+            if (InterlockedCompareExchangeAcquire((LONG*)&m_SyncBlockValue, newValue, oldValue) == oldValue)
+#endif
+            {
+                return HeaderLockResult::Success;
+            }
+
+            // Someone else just beat us to the lock.
+            // Try again.
+            YieldProcessorNormalized(i);
+            continue;
+        }
+
+        if (oldValue & BIT_SBLK_IS_HASH_OR_SYNCBLKINDEX)
+        {
+            return HeaderLockResult::UseSlowPath;
+        }
+
+        // The header is transitioning - use the slow path
+        if (oldValue & BIT_SBLK_SPIN_LOCK)
+        {
+            return HeaderLockResult::UseSlowPath;
+        }
+
+        // Here we know we have the "thin lock" layout, but the lock is not free.
+        // It could still be the recursion case - compare the thread id to check
+        if (tid != (DWORD)(oldValue & SBLK_MASK_LOCK_THREADID))
+        {
+            // Someone else owns the lock.
+            // Try again.
+            YieldProcessorNormalized(i);
```

Comment:
The managed equivalent polled for GC. This one is less friendly to GC suspension.